egressSafelist: Allow specifying different ports than 80 and 443

issyl0 opened the pull request:
- We tried to add a rule for RDS access, but the hardcoded ports were 80
  and 443. RDS uses 5432.
- This refactors the entire Helm chart for specifying service entries to
  account for all the different things tenants could want: they now
  specify all the relevant fields in their `values.yaml` in the
  `cluster-config` repos.
- Now we support things other than HTTP and HTTPS, name the chart
  `egress-safelist.yaml` rather than `https-egress-safelist.yaml`. This
  means that we can delete `http-egress-safelist.yaml`.

Co-authored-by: Issy Long <isabell.long@digital.cabinet-office.gov.uk>
Co-authored-by: Chris Farmiloe <chris.farmiloe@digital.cabinet-office.gov.uk>

https://trello.com/c/uEbwPqoN/111-make-gsp-service-operator-postgres-support-work

issyl0 committed:
Allow specifying different ports than 80 and 443

- We tried to add a rule for RDS access, but the hardcoded ports were 80
  and 443. RDS uses 5432.
- This refactors the entire Helm chart for specifying service entries to
  account for all the different things tenants could want: they now
  specify all the relevant fields in their `values.yaml` in the
  `cluster-config` repos.
- Now we support things other than HTTP and HTTPS, name the chart
  `egress-safelist.yaml` rather than `https-egress-safelist.yaml`. This
  means that we can delete `http-egress-safelist.yaml`.

Co-authored-by: Issy Long <isabell.long@digital.cabinet-office.gov.uk>
Co-authored-by: Chris Farmiloe <chris.farmiloe@digital.cabinet-office.gov.uk>

Krenair commented on charts/gsp-cluster/templates/02-gsp-system/egress-safelist.yaml:
```diff
@@ -1,22 +1,14 @@
   namespace: {{ $.Release.Namespace }}
   labels:
-    app.kubernetes.io/name: {{ include "gsp-cluster.name" $ }}-http-egress-safelist-{{ $egressFqdn.name }}
+    app.kubernetes.io/name: {{ include "gsp-cluster.name" $ }}-https-egress-safelist-{{ $i }}
```
```suggestion
    app.kubernetes.io/name: {{ include "gsp-cluster.name" $ }}-egress-safelist-{{ $i }}
```

issyl0 commented on charts/gsp-cluster/templates/02-gsp-system/egress-safelist.yaml:
```diff
@@ -1,22 +1,14 @@
-{{- range $egressFqdn := .Values.httpEgressSafelist }}
+{{- range $i, $egress := .Values.serviceEntry }}
```
```suggestion
{{- range $i, $egress := .Values.egressSafelist }}
```

issyl0 committed:
Get the lint-values script matching on the new `egressSafelist` syntax

Krenair commented on charts/gsp-cluster/templates/02-gsp-system/egress-safelist.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: networking.istio.io/v1alpha3
+kind: ServiceEntry
+metadata:
+  name: {{ $.Release.Name }}-egress-safelist-{{ $i }}
```
Naming based on the index feels weird and potentially dangerous if someone adds something mid way through the list, or removes anything at all

chrisfarms replied:
Yeah, guess our options are either using a map:

```
egressSafelist:
  integrationHub:
    hosts: ["www.integration.signin.service.gov.uk"]
    ports:
    - name: verify-integration-https
      number: 443
      protocol: TCP
    location: MESH_EXTERNAL
    resolution: DNS
  productionHub:
    hosts: ["www.signin.service.gov.uk"]
    ports:
    - name: verify-https
      number: 443
      protocol: TCP
    location: MESH_EXTERNAL
    resolution: DNS
```

or adding the `name` back

```
egressSafelist:
- name: integration-hub
  service:
    hosts: ["www.integration.signin.service.gov.uk"]
    ports:
    - name: verify-integration-https
      number: 443
      protocol: TCP
    location: MESH_EXTERNAL
    resolution: DNS
- name: production-hub
  service:
    hosts: ["www.signin.service.gov.uk"]
    ports:
    - name: verify-https
      number: 443
      protocol: TCP
    location: MESH_EXTERNAL
    resolution: DNS
```

issyl0 replied:
Resolved in 1f930af.

Krenair reviewed (changes requested):
.

blairboy362 commented:
Is this an opportunity, given that the safelist now only adds `ServiceEntry` resources, to move this templating out and get "tenants" to create them (in say `*-main` via the `configure-namespace` mechanic) explicitly? It might be easier to reason about and removes some complexity from the GSP...

Just my 2 cents.

chrisfarms commented:
@blairboy362  I think I'd prefur to continue using the gsp-cluster chart for this (for now) without introducing a new special case, since I think it's easier to say "global cluster config stuff is in the cluster-config values.yaml" ... but I don't disagree that we may want to completely seperate cluster-config from gsp-cluster

issyl0 committed:
Reinstate naming based on the name from values.yaml

- Naming the ServiceEntry based on an index was brittle if someone added
  a block to values.yaml in the middle instead of at the end of the
  existing egressSafelist entries.
